File: cocotb/tests/user_project/user_address_space.c
#include <common.h>



// --------------------------------------------------------

void main()
{
    enable_user_interface();
    // first 3 addresses 
    (*(volatile unsigned int*) (USER_SPACE_ADDR    )) = 0x97cf0d2d; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR +4 )) = 0xbc748313; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR +8 )) = 0xbfda8146; 

    // last 3 addresses 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + USER_SPACE_SIZE -8 )) = 0x5f5e36b1; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + USER_SPACE_SIZE -4 )) = 0x0c1fe9d9; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + USER_SPACE_SIZE    )) = 0x6d12d2b8; 


    // random addresses inside the user space 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x72C  )) = 0xdcd244d1; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x41198)) = 0x0da37088; 
    (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x7770 )) = 0x7b8e4345; 
    // random read 
    int temp;
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x9F44  )); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x58    )); 
    temp = (*(volatile unsigned int*) (USER_SPACE_ADDR + 0x3602EC)); 

    // addresses outside user space - injecting error if user project ack is affected
    configure_gpio(14,GPIO_MODE_MGMT_STD_OUTPUT);
    configure_gpio(15,GPIO_MODE_MGMT_STD_OUTPUT);
    
    // finish with writing last address with Fs
    (*(volatile unsigned int*)(USER_SPACE_ADDR + USER_SPACE_SIZE)) = 0xFFFFFFFF; 
    

}

File: cocotb/tests/user_project/user_address_space.py
import random
import cocotb
from cocotb.triggers import FallingEdge, RisingEdge, ClockCycles, Timer, Edge
import cocotb.log
from interfaces.cpu import RiskV
from interfaces.defsParser import Regs
from cocotb.result import TestSuccess
from tests.common_functions.test_functions import *

from interfaces.caravel import GPIO_MODE
from interfaces.common import Macros


reg = Regs()


@cocotb.test()
@repot_test
async def user_address_space(dut):
    caravelEnv = await test_configure(dut, timeout_cycles=37599)
    cocotb.log.info(f"[TEST] Start user_address_space test")
    ack_hdl = caravelEnv.caravel_hdl.mprj.addr_space_testing.wbs_ack_o
    addr_hdl = caravelEnv.caravel_hdl.mprj.addr_space_testing.addr
    data_hdl = caravelEnv.caravel_hdl.mprj.addr_space_testing.data
    start_addr = Macros["USER_SPACE_ADDR"]
    print(f"user space adddress = {start_addr}")
    user_size = Macros["USER_SPACE_SIZE"]
    addr_arr = (
        start_addr,
        start_addr + 4,
        start_addr + 8,
        start_addr + user_size - 8,
        start_addr + user_size - 4,
        start_addr + user_size,
        start_addr + 0x72C,
        start_addr + 0x41198,
        start_addr + 0x7770,
        start_addr + 0x9F44,
        start_addr + 0x58,
        start_addr + 0x3602EC,
        start_addr + user_size,
    )
    data_arr = (
        0x97CF0D2D,
        0xBC748313,
        0xBFDA8146,
        0x5F5E36B1,
        0x0C1FE9D9,
        0x6D12D2B8,
        0xDCD244D1,
        0x0DA37088,
        0x7B8E4345,
        0x00000777,
        0x00000777,
        0x00000777,
        0xFFFFFFFF,
    )
    print([hex(i) for i in addr_arr])
    for addr, data in zip(addr_arr, data_arr):
        await RisingEdge(ack_hdl)
        if addr_hdl.value.integer != addr:
            cocotb.log.error(
                f"[TEST] seeing unexpected address {hex(addr_hdl.value.integer)} expected {hex(addr)}"
            )
        elif data_hdl.value.integer != data:
            cocotb.log.error(
                f"[TEST] seeing unexpected data {hex(data_hdl.value.integer)} expected {hex(data)} address {hex(addr)}"
            )
        else:
            cocotb.log.info(
                f"[TEST] seeing the correct data {hex(data)} from address {hex(addr)}"
            )
